lib/SILOptimizer/Transforms/AccessEnforcementDom.cpp

```diff
@@ -0,0 +1,257 @@
+    // we can't eliminate dominated checks even when we can prove that
+    // the dominated scope has no internal nested conflicts.
+    keypathEntries.insert(fullApply.getInstruction());
+  } else if (auto *BUAI = dyn_cast<BeginUnpairedAccessInst>(instr)) {
+    // We have an Implementation that handles this in the analyzer
+    // and optimizer in a previous commit, However,
+    // In order to simplify the code, at least in the initial version,
+    // We decided to just bail on these extremely rare cases
+    return false;
+  }
+  return true;
+}
+
+void DominatedAccessRemoval::visitBeginAccess(
+    BeginAccessInst *beginAccess, AccessedStorage storage,
+    AccessedStorageInfo &visitedDomAccessesToStorageInfo) {
+  if (!storage.isUniquelyIdentifiedOrClass()) {
+    // No cannot do anything about this location -
+    // bail on trying to turn it to static
+    return;
+  }
+
+  // checks if the current storage has never been seen before
+  auto predNewStorageLoc = [&](AccessedStoragePair it) {
+    auto currStorage = it.second;
+    return !currStorage.isDistinctFrom(storage);
+  };
+  // The size of this list should be tiny: number of storage locations
+  auto visitedAccessesIt =
+      std::find_if(visitedDomAccessesToStorageInfo.begin(),
+                   visitedDomAccessesToStorageInfo.end(), predNewStorageLoc);
+
+  if (visitedAccessesIt == visitedDomAccessesToStorageInfo.end()) {
+    // We've never seen this one before - just add it and return
+    visitedDomAccessesToStorageInfo.push_back(
+        std::make_pair(beginAccess, storage));
+    return;
+  }
+
+  // If the currnet access has nested conflict,
+  // we can't remove it by finding a dominating access
+  if (!beginAccess->hasNoNestedConflict()) {
+    return;
+  }
+
+  BeginAccessInst *parentBeginAccess = visitedAccessesIt->first;
+  domPairs.push_back(std::make_pair(parentBeginAccess, beginAccess));
+}
+
+bool DominatedAccessRemoval::analyzeDomSubTree(
+    SILBasicBlock *block,
+    AccessedStorageInfo &visitedDomAccessesToStorageInfo) {
+  // We will modify the incoming visitedDomAccessesToStorageInfo,
+  // after finishing with the sub-tree,
+  // we need to restore it to its previous state.
+  // the state should be small because the number of unique storage
+  // locations, which is the size of the data structure,
+  // should be quite small. a handful at most.
+  auto numOfElems = visitedDomAccessesToStorageInfo.size();
+
+  // analyze the current block:
+  for (auto &instr : *block) {
+    if (!visitInstruction(&instr, visitedDomAccessesToStorageInfo))
+      return false;
+  }
+  // do the same for each child:
+  auto *domNode = domInfo->getNode(block);
+  for (auto *child : *domNode) {
+    if (!analyzeDomSubTree(child->getBlock(), visitedDomAccessesToStorageInfo))
+      return false;
+  }
+
+  // Restore the sets to their previous state as described above,
+  // removing all "new" elements
+  auto newNumOfElems = visitedDomAccessesToStorageInfo.size();
```

Comment:
please inline this expression into the assert. Otherwise you'll get a warning in the release build